internal/server/request_start_middleware.go

```diff
@@ -0,0 +1,29 @@
+package server
+
+import (
+	"net/http"
+	"strconv"
+	"time"
+)
+
+type RequestStartMiddleware struct {
+	next http.Handler
+}
+
+func WithRequestStartMiddleware(next http.Handler) http.Handler {
+	return &RequestStartMiddleware{
+		next: next,
+	}
+}
+
+func (h *RequestStartMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Header.Get("X-Request-Start") == "" {
+		r.Header.Set("X-Request-Start", h.nowStr())
+	}
```

Comment:
Add this at the top:

``` go
const (
	requestStartHeader = "X-Request-Start"
)
```

```suggestion
	if r.Header.Get(requestStartHeader) == "" {
		r.Header.Set(requestStartHeader, h.nowStr())
	}
```

You can then also reuse this in tests.